tests/L0/run_transformer/test_fused_rope.py

```diff
@@ -0,0 +1,138 @@
+"""Test for fused RoPE functions.
+
+Ref: https://github.com/NVIDIA/Megatron-LM/blob/40becfc96c4144985458ac0e0fae45dbb111fbd2/megatron/fused_kernels/tests/test_fused_kernels.py
+"""  # NOQA
+import itertools
+from typing import Tuple, Union
+
+import torch
+from torch.testing._internal import common_utils
+
+
+class FusedRoPEFunc(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx, t: torch.Tensor, cos_: torch.Tensor, sin_: torch.Tensor
+    ) -> torch.Tensor:
+        import fused_rotary_positional_embedding
+
+        output = fused_rotary_positional_embedding.forward(t, cos_, sin_)
+        ctx.save_for_backward(cos_, sin_)
+
+        return output
+
+    @staticmethod
+    def backward(
+        ctx, grad_output: torch.Tensor
+    ) -> Tuple[Union[torch.Tensor, None], ...]:
+        import fused_rotary_positional_embedding
+
+        cos_, sin_ = ctx.saved_tensors
+        grad_q = fused_rotary_positional_embedding.backward(grad_output, cos_, sin_)
+
+        return grad_q, None, None
+
+
+def apply_rotary_pos_emb_fused(t: torch.Tensor, freqs: torch.Tensor) -> torch.Tensor:
+    cos_ = torch.cos(freqs).to(t.dtype)
+    sin_ = torch.sin(freqs).to(t.dtype)
+    return FusedRoPEFunc.apply(t, cos_, sin_)
```

Comment:
out of curiosity, wouldn't it be useful to have this in `apex.transformer.functiona` namespace?

Reply:
Thanks for your review. I have added two functions to the `apex.transformer.functional` namespace.
1. `fused_apply_rotary_pos_emb`, which is a drop-in replacement for the current `apply_rotary_pos_emb` in Megatron Core.
2. `fused_apply_rotary_pos_emb_cached`, which would be beneficial when MCore implements caching for the rotary positional embedding.